File: .github/scripts/check-markdown-links.test.py
#!/usr/bin/env python3
from __future__ import annotations

import importlib.util
import tempfile
import unittest
from pathlib import Path

SCRIPT = Path(__file__).with_name("check-markdown-links.py")
SPEC = importlib.util.spec_from_file_location("check_markdown_links", SCRIPT)
assert SPEC is not None and SPEC.loader is not None
checker = importlib.util.module_from_spec(SPEC)
SPEC.loader.exec_module(checker)


class MarkdownLinkCheckerTest(unittest.TestCase):
    def test_markdown_target_parsing_policy(self) -> None:
        cases = [
            (
                "inline relative link",
                "See [guide](guide.md).",
                [(1, "guide.md")],
            ),
            (
                "reference-style definition",
                "See [guide][g].\n\n[g]: guide.md",
                [(3, "guide.md")],
            ),
            (
                "inline code historical path is ignored",
                "Historical path: `[removed](docs/removed.md)`.",
                [],
            ),
            (
                "fenced code historical path is ignored",
                "```text\n[removed](docs/removed.md)\n```",
                [],
            ),
            (
                "external link is still parsed before locality filtering",
                "[OpenAI](https://openai.com/)",
                [(1, "https://openai.com/")],
            ),
        ]

        for name, text, expected in cases:
            with self.subTest(name=name):
                self.assertEqual(expected, list(checker.markdown_targets(text)))

    def test_local_candidate_policy(self) -> None:
        with tempfile.TemporaryDirectory() as temp_dir:
            repo = Path(temp_dir).resolve()
            docs = repo / "docs"
            docs.mkdir()
            source = docs / "README.md"
            source.write_text("# Docs\n", encoding="utf-8")
            guide = docs / "guide.md"
            guide.write_text("# Guide\n", encoding="utf-8")

            cases = [
                (
                    "valid relative repository link",
                    "guide.md",
                    guide,
                ),
                (
                    "broken relative repository link",
                    "missing.md",
                    docs / "missing.md",
                ),
                (
                    "root repository link",
                    "/docs/guide.md",
                    guide,
                ),
                (
                    "root web route",
                    "/api/health",
                    None,
                ),
                (
                    "external URL",
                    "https://example.com/docs",
                    None,
                ),
                (
                    "in-page anchor",
                    "#section",
                    None,
                ),
            ]

            for name, target, expected in cases:
                with self.subTest(name=name):
                    self.assertEqual(expected, checker.local_candidate(repo, source, target))

            self.assertTrue(checker.local_candidate(repo, source, "guide.md").exists())
            self.assertFalse(checker.local_candidate(repo, source, "missing.md").exists())


if __name__ == "__main__":
    unittest.main()

File: .github/scripts/check-markdown-links.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { check, localCandidate, markdownTargets } from './check-markdown-links.mjs';

test('parses inline and reference links while ignoring code', () => {
  assert.deepEqual(markdownTargets('See [guide](guide.md).\n[g]: <docs/a.md>\n`[x](no.md)`\n```\n[x](no.md)\n```'), [[1, 'guide.md'], [2, 'docs/a.md']]);
});

test('resolves relative, root-style, external, anchor, and route targets', (t) => {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'markdown-links-'));
  t.after(() => fs.rmSync(root, { recursive: true, force: true }));
  fs.mkdirSync(path.join(root, 'docs'));
  const source = path.join(root, 'docs', 'README.md');
  fs.writeFileSync(source, '# Docs\n');
  fs.writeFileSync(path.join(root, 'docs', 'guide.md'), '# Guide\n');
  assert.equal(localCandidate(root, source, 'guide.md'), path.join(root, 'docs', 'guide.md'));
  assert.equal(localCandidate(root, source, '/docs/guide.md'), path.join(root, 'docs', 'guide.md'));
  assert.equal(localCandidate(root, source, '/api/health'), null);
  assert.equal(localCandidate(root, source, 'https://example.com/'), null);
  assert.equal(localCandidate(root, source, '#section'), null);
});

test('checker reports broken links and ignores external/root web routes', (t) => {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'markdown-links-'));
  t.after(() => fs.rmSync(root, { recursive: true, force: true }));
  fs.mkdirSync(path.join(root, 'docs'));
  fs.writeFileSync(path.join(root, 'docs', 'README.md'), '[bad](missing.md)\n[route](/api/health)\n[external](https://example.com)\n');
  assert.deepEqual(check(root), ["docs/README.md:1: broken local link 'missing.md' -> docs/missing.md"]);
});

test('preserves parent-relative segments, URL decoding, and query/fragment handling', (t) => {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'markdown-links-'));
  t.after(() => fs.rmSync(root, { recursive: true, force: true }));
  fs.mkdirSync(path.join(root, 'docs', 'nested'), { recursive: true });
  fs.writeFileSync(path.join(root, 'docs', 'guide with space.md'), '# Guide\n');
  const source = path.join(root, 'docs', 'nested', 'README.md');
  fs.writeFileSync(source, '[parent](../guide%20with%20space.md?view=1#section)\n');
  assert.deepEqual(check(root), []);
});
